cdap/resource_streaming_program.go

```diff
@@ -0,0 +1,248 @@
+		Read:   resourceStreamingProgramRead,
+		Delete: resourceStreamingProgramDelete,
+		Exists: resourceStreamingProgramExists,
+
+		Schema: map[string]*schema.Schema{
+			"namespace": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of the namespace in which this resource belongs. If not provided, the default namespace is used.",
+				DefaultFunc: func() (interface{}, error) {
+					return defaultNamespace, nil
+				},
+			},
+			"app": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the application.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the program.",
+				DefaultFunc: func() (interface{}, error) {
+					return "DataStreamsSparkStreaming", nil
+				},
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "One of flows, mapreduce, services, spark, workers, or workflows.",
+				ValidateFunc: validation.StringInSlice([]string{"flows", "mapreduce", "services", "spark", "workers", "workflows"}, false),
+				DefaultFunc: func() (interface{}, error) {
+					return "spark", nil
+				},
+			},
+			"runtime_arguments": {
+				Type:        schema.TypeMap,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The runtime arguments used to start the program",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceStreamingProgramCreate(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+	name := d.Get("name").(string)
+
+	addr := urlJoin(
+		config.host,
+		"/v3/namespaces", d.Get("namespace").(string),
+		"/apps", d.Get("app").(string), d.Get("type").(string),
+		name)
+	startAddr := urlJoin(
+		addr, "start")
+	statusAddr := urlJoin(
+		addr, "status")
+
+	argsObj := make(map[string]interface{})
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	for {
+		p, err := getProgramStatus(config, statusAddr)
+		if err != nil {
+			return err
+		}
+
+		switch s := p.Status; s {
+		case "PROVISIONING":
+			log.Println("program still in PROVISIONING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		case "STARTING":
+			log.Println("program still in STARTING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		case "STOPPED":
+			log.Println("program still in STOPPED state, waiting 10 seconds. This may occur when redeploying a previously deployed pipeline.")
+			time.Sleep(10 * time.Second)
+		case "RUNNING":
+			log.Println("program successfully reached RUNNING state.")
+			d.SetId(d.Get("app").(string))
```

Comment:
I think we should set the ID sooner, main reason is if the program is created but the wait fails then currently Terraform will think the program was not created.

If you set the ID earlier but the wait fails, the resource becomes tainted and the delete func gets called first instead.

We should also set the program name to be the ID instead of the app. If we use the app then if you have two streaming programs on the same app it will conflict and fail.

Reply:
Actually nevermind, the way you did it right now makes sense. If the state of the resource never reaches STARTED then we can't assume it was created. If it failed then it should not be set and if it's anything else its in progress so keep retrying... ok.

Reply:
Or wait... FAILED means failed to *start*? If so then yeah the ID should be set when the resource is physically created (which the earlier call should have done).